Implement a `NIOAsyncWriter`

FranzBusch opened the pull request:
# Motivation
We previously added the `NIOAsyncSequenceProducer` to bridge between the NIO channel pipeline and the asynchronous world. However, we still need something to bridge writes from the asynchronous world back to the NIO channel pipeline.

# Modification
This PR adds a new `NIOAsyncWriter` type that allows us to asynchronously `yield` elements to it. On the other side, we can register a `NIOAsyncWriterDelegate` which will get informed about any written elements. Furthermore, the synchronous side can toggle the writability of the `AsyncWriter` which allows it to implement flow control.
A main goal of this type is to be as performant as possible. To achieve this I did the following things:
- Make everything generic and inlinable
- Use a class with a lock instead of an actor
- Provide methods to yield a sequence of things which allows users to reduce the amount of times the lock gets acquired.

# Result
We now have the means to bridge writes from the asynchronous world to the synchronous

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+//
+//===----------------------------------------------------------------------===//
+
+import Atomics
+import NIOConcurrencyHelpers
```
Can we move these into the compiler guards?

FranzBusch replied:
We can but we are not aligned on this. Even in my previous `NIOAsyncSequenceProducer` PR we haven't done this.

glbrntt replied:
IMHO if it isn't required it shouldn't be imported: if we don't have Swift > 5.5.2 then we shouldn't import these

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+
+    /// This method is called once the ``NIOAsyncWriter`` is terminated.
+    ///
+    /// Termination happens if:
```
Is this always called exactly once?

FranzBusch replied:
Yes this is guaranteed to be called exactly one same as the `didTerminate` on the `NIOAsyncSeuqenceProducerDelegate`

glbrntt replied:
Can we document that?

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+    case failure(Failure)
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+/// Errors thrown by the ``NIOAsyncWriter``.
```
```suggestion
/// Errors thrown by the ``NIOAsyncWriter``.
@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
```

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+>: Sendable where Delegate.Element == Element, Delegate.Failure == Failure {
+    /// Simple struct for the return type of ``NIOAsyncWriter/makeWriter(elementType:failureType:isWritable:delegate:)``.
+    ///
+    /// This struct contains two properties:
+    /// 1. The ``sink`` which should be retained by the consumer and is used to toggle the writability.
+    /// 2. The ``writer`` which is the actual ``NIOAsyncWriter`` and should be passed to the producer.
```
Can we move the docs onto the properties?

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+    ///
+    /// If we get move-only types we should be able to drop this class and use the `deinit` of the ``NIOAsyncWriter`` struct itself.
+    ///
+    /// - Important: This is safe to be unchecked ``Sendable`` since the `storage` is ``Sendable`` and `immutable`.
+    @usableFromInline
+    /* fileprivate */ internal final class InternalClass: @unchecked Sendable {
```
If `Storage` is `Sendable` and immutable then this doesn't need to be `@unchecked`.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+    }
+
+    @usableFromInline
+    /* private */ internal let internalClass: InternalClass
```
nit: we generally prefix with an underscore for private-but-usable-from-inline properties.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+    /// If the ``NIOAsyncWriter`` was writable when the sequence was yielded, the sequence will be forwarded
+    /// right away to the delegate. If the ``NIOAsyncWriter`` was _NOT_ writable then the sequence will be buffered
+    /// until the ``NIOAsyncWriter`` becomes writable again.
+    func didYield<S: Sequence>(contentsOf sequence: S) where S.Element == Element
```
Do we always yield a complete sequence?

FranzBusch replied:
We always `yield` the complete sequence that the user yielded to us. However, the user might have also yielded a single element which we will yield as a `CollectionOfOne` sequence.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+    /// - Parameter completion: A ``NIOAsyncWriterCompletion`` which indicates whether publishing has finished normally
+    /// or failed with an error.
+    @inlinable
+    public func finish(completion: NIOAsyncWriterCompletion<Failure>) {
```
Is it actually worth having a _public_ `NIOAsyncWriterCompletion` here rather than say, `finish()` and `finish(error:)`?

`finish(completion: .finished)` and `finish(completion: .failure(.someError))` seem unnecessarily verbose.

FranzBusch replied:
I adopted the spelling from `Combine`. We also have the spelling from `AsyncThrowingStream.Continuation` which has `finish(throwing error: Failure? = nil`.
I like your approach the best though having to separate methods

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+    /// The underlying sink of the ``NIOAsyncWriter``. This type allows to toggle the writability of the ``NIOAsyncWriter``.
+    public struct Sink {
+        @usableFromInline
+        /* private */ internal var storage: Storage
```
is this intentionally `var`?

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+        ///
+        /// If the writer becomes writable again all suspended yields will be resumed and the produced elements will be forwarded via
+        /// the ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` method.
+        public func toggleWritability() {
```
Is this the right API? I wonder if setting the writability explicitly would be more useful (if anything it makes the tests more obvious). In `actor` versions it _had_ to be toggle because calls to set the writability could be reordered.

FranzBusch replied:
I kept this in because to get the writability of a `Channel` you need to load the value from the underlying atomic. Since, that load is normally done using a `relaxed` load I wasn't sure if this is safe to do in the `channelWritabilityChanged` method. However, after talking with @Lukasa this is safe to do so I am changed the API to be `setWritability(to:)`.

glbrntt reviewed (commented):
Only just got to the meat of this PR; will pick up the rest of the review later. Leaving my comments so far.

FranzBusch commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,750 @@
+                self.lock.unlock()
+
+                delegate?.didYield(contentsOf: suspendedYields.lazy.map { $0.elements }.flatMap { $0 })
+                suspendedYields.forEach { $0.continuation.resume() }
```
I prefer keeping the `forEach` since it is more concise for a very simple task. However, in general I agree to prefer `for ...in` over `forEach` for anything that is more complex

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,763 @@
+                    // or that the call to `yield` is coming afterwards. We need to store
+                    // the ID here. However, if the yield already happened we will never remove the
+                    // stored ID. The only way to avoid doing this would be storing every ID
+                    cancelledYields.append(yieldID)
```
Do we need to be `.modifying` here?

FranzBusch replied:
Good catch!

FranzBusch commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,763 @@
+        }
+
+        @inlinable
+        /* fileprivate */ internal mutating func yield<S: Sequence>(
```
I see. I kinda followed what I did for the `NIOAsyncSequenceProducer` where I have `next()` and `next(for:)`. If you want I can change it. Up to you!

glbrntt replied:
Don't feel strongly about it, a small comment would help though.

FranzBusch commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,763 @@
+        /// subsequent calls to ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` will suspend.
+        ///
+        /// - Parameter writability: The new writability of the ``NIOAsyncWriter``.
+        public func setWritability(to writability: Bool) {
```
Hm, I am not 100% convinced. If we change it to a property we also have to provide a `getter` which we would need to go through the lock to get the actual value. Since, I don't expect anyone to be actually querying for the writable of the `Sink` I am kinda hesitant to make that change.

glbrntt replied:
You're right, we don't need a getter here.

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+/// this type has three generic parameters where at least two should be known statically and it is really awkward to spell out this type.
+/// Moreover, having a wrapping type allows to optimize this to specialized calls if all generic types are known.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncWriter<
```
I'm not sure, I'm a fan of this being a struct. The AsyncWriter has clearly reference semantics.

FranzBusch replied:
Fair point, I think making this a class isn't bad. I went with the struct first approach but open to change it

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+
+/// The delegate of the ``NIOAsyncWriter``.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public protocol NIOAsyncWriterDelegate: Sendable {
```
Is there any situation, in which this could be a `struct`? Can we require this to inherit from `AnyObject`?

FranzBusch replied:
Yeah sometimes you want to use a struct for this to have an intermediate type that is implementing this. I think forcing this to inherit from `AnyObject` is overly restrictive.

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// The underlying sink of the ``NIOAsyncWriter``. This type allows to toggle the writability of the ``NIOAsyncWriter``.
+    public struct Sink {
```
What's the benefit of this type? Ensure that a consumer of the `NIOAsyncWriter` can only call `setWritability` on it? In my mind the `OutboundChannelHandler` implements the Delegate protocol and holds a reference to the writer. `OutboundChannelHandler` can then easily toggle writability on the `NIOAsyncWriter`.

FranzBusch replied:
The benefit of this type is that the producer and the delegate are holding two separate types. This allows us to make assumption of when the `NIOAsyncWriter` actually gets deallocated since this means we can tell the delegate that we terminated. I stated in the doc comments of `makeWriter` that you should hold the `Sink` in the place where you consume the writes and the `AsyncWriter` in the async world.

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+        @usableFromInline
+        /* private */ internal let _yieldIDCounter = ManagedAtomic<UInt64>(0)
+        /// The state machine.
+        @usableFromInline
+        /* private */ internal var _stateMachine: StateMachine
+        /// The delegate.
+        @usableFromInline
+        /* private */ internal var _delegate: Delegate?
```
I assume, holding the `Delegate`, or not holding the `Delegate` is part of the internal state. Should the delegate be referenced from the StateMachine? Then the stateMachine should return the Delegate when needed for some actions. This way, we should get rid of a bunch of `?`.

FranzBusch replied:
Yeah happy to change this. Especially now that we are calling the delegate while holding the lock

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        @usableFromInline
+        /* private */ internal let _lock = Lock()
```
What does the underscore mean here?

FranzBusch replied:
The underscore her is due to the fact that we would like make the lock `private` but can't since we want to add `@usableFromInline`.

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+            var yieldID: UInt64
+            /// The yield's produced sequence of elements.
+            @usableFromInline
+            var elements: AnySequence<Element>
```
Should the AsyncWriter be generic over the Sequence type as well? How often do user's want to send different Sequence types? For the SequenceOfOne/Two case users could always opt for an enum that implements Sequence.

FranzBusch replied:
Yeah, I didn't like the point of holding an `AnySequence` here. If we do that, then we should get rid of the yielding of a single element.

fabianfett replied:
> If we do that, then we should get rid of the yielding of a single element.
I think we must and I consider this a good thing :P

glbrntt replied:
> Should the AsyncWriter be generic over the Sequence type as well? How often do user's want to send different Sequence types?

I'm not sure that's a good idea in practice. I think the typical usage pattern would be for a framework to create a writer and pass it to a caller. That would mean the framework would be responsible for choosing what the sequence type is _or_ the caller has to ask the framework for a writer with a given sequence type (which is a pretty awkward API).

> For the SequenceOfOne/Two case users could always opt for an enum that implements Sequence.

Not offering a yield-one API and requiring users to implement an enum like this makes for a really user hostile API IMHO.

---

FWIW I also don't like `AnySequence` here but I think we should keep it for the time being (it is an implementation detail and can be changed later). 

One alternative would be to always store `[Element]` and special case the API to avoid copies when an array is provided and copy into an array when anything else is provided. 

Taking that further we could be to pull a 'NIOAny' trick and have an enum of `array`, `one`, and `anySequence`.

That said, I think we should focus on getting the API right first and then compare perf.

FranzBusch replied:
I agree with what George said. Making the whole thing generic of the sequence that is yielded severely limits the usage of the type and makes it very awkward to hold. I would stick with `AnySequence` for now and as George said we can always make it faster by applying some of our tricks.

Lukasa replied:
As discussed offline, I think we should change this to remove the `AnySequence`.

fabianfett commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+        /* fileprivate */ internal func yield<S: Sequence>(contentsOf sequence: S) async throws where S.Element == Element {
+            // Using relaxed is fine here since we do not need any strict ordering just a
+            // unique ID for every yield.
+            let yieldID = self._yieldIDCounter.loadThenWrappingIncrement(ordering: .relaxed)
+
+            try await withTaskCancellationHandler {
+                // We are manually locking here to hold the lock across the withCheckedContinuation call
+                self._lock.lock()
+
+                let action = self._stateMachine.yield(yieldID: yieldID)
```
What's the benefit of having the `yieldIDCounter` as an atomic, if we acquire a lock afterwards anyway?

fabianfett reviewed (commented):
Great start. Left some broader questions for now.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,763 @@
+    @usableFromInline
+    let _code: _Code
+
+    public static let alreadyFinished: Self = .init(_code: .alreadyFinished)
```
Up to you. This PR is already pretty large. I'm happy to leave it for now.

glbrntt replied:
This needs to become `func` now so we capture the file and line it was called, otherwise it'll always be this file and line

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,789 @@
+    var line: Int
+
+    @inlinable
+    init(_code: _Code, file: String = #fileID, line: Int = #line) {
```
We should remove the defaults here so we ensure they're always provided by the helpers

Lukasa reviewed (changes requested):
I think we're mishandling the delegate throughout this code.

Because we only call the delegate from _outside_ the lock, we do not guarantee to maintain ordering of yields. We can suspend immediately after we drop the lock, and then have a completely unrelated Task come in and yield results, get an immediate call to `didYield`, and then drop into the delegate, beating the first Task.

This didn't matter in the previous delegate, which was only being called with signals that were themselves well-ordered, but it does matter here. We need a fundamentally different behaviour that maintains ordering, probably by way of keeping the lock held across these calls. That will require some thought as to how it interacts with `setWritability`, which we need to be careful wtih.

glbrntt reviewed (approved):
Looks great! Left a few documentation nits/fixes.

FranzBusch commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,823 @@
+    /// The `Element` type of the delegate and the writer.
+    associatedtype Element: Sendable
+    /// The `Failure` type of the delegate and the writer.
+    associatedtype Failure: Error = Never
```
AFAIK, they existed for quite some time; however, there are various quirks in the current type checker that require you to still explicitly specify them in the implementation which is kinda sad.

glbrntt replied:
Seems to be an undocumented feature: https://github.com/apple/swift/issues/51269

Lukasa commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,851 @@
+        /* private */ internal let _lock = Lock()
+        /// The counter used to assign an ID to all our yields.
+        @usableFromInline
+        /* private */ internal let _yieldIDCounter = ManagedAtomic<UInt64>(0)
```
Can we wrap this in a `struct` that gives us back an object that isn't a `UInt64` that has the API surface we want? The way we use this made me think for a while that it was supposed to be a monotonically increasing value, but in actuality it just needs to be a _unique_ value. Having a type that encapsulates this logic and doesn't encourage comparison would be helpful.

FranzBusch commented:
@Lukasa I changed the API to always yield a `Deque` now and I added a performance test for this. Let me know what you think!

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,934 @@
+    /// - Note: Calling this function more than once has no effect.
+    /// - Parameter failure: The failure indicating why the writer finished.
+    @inlinable
+    public func finish(with failure: Failure) {
```
Minor bikeshedding... I don't love the use of `with` here. To me, `finish(with:)` suggests a completion value (rather than completing as a result of a failure).

Some previous art:
- `AsyncThrowingStream.Continuation` has `finish(throwing:)`
- `AsyncThrowingChannel` has `fail(_:)`

I prefer `finish(throwing:)` so that it's similar to `finish()` over `fail(_:)`.

glbrntt replied:
Agree `throwing:` is weird if it isn't thrown. It does at least make it obvious when reading that it's an error. I think `finish(failure:)`/`finish(error:)` would be fine.

FranzBusch commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,934 @@
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncWriter<
+    Element,
+    Failure,
```
They already are. The delegate is currently restricting the `Element` to be `Sendable` which propagates to this generic type through the constraints. Furthermore, `Failure` is restricted to `Error` which inherits from `Sendable`.

FranzBusch replied:
Actually taking this back since it produces a warning since it is a duplicate constraint.

FranzBusch commented:
@Lukasa @glbrntt @gjcairo I just pushed a new commit that changes a few things after our offline discussion:
- Allow sending an `Error` with the `Sink/finish(error:)` method. Any suspended or subsequent call to `NIOAsyncWriter/yield` will throw this error. It provides a way to surface an underlying error to the callers of the `NIOAsyncWriter`. We have been doing this in `grpc-swift` to surface errors from the channel pipeline 
- I removed the generic `Failure` parameter from the `NIOAsyncWriter` since we agreed that there is little value in providing a generic here over just an `Error`. Furthermore, since we want to allow to send an `Error` from the `Sink/finish(error:)` method it would mean we would have to introduce another generic parameter.
- Added a few more tests for the new state and updated the documentation.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,1060 @@
+    /// becomes writable again. If the calling `Task` gets cancelled at any point the call to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// will be resumed.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
```
I don't think this is accurate.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,1060 @@
+    /// becomes writable again. If the calling `Task` gets cancelled at any point the call to ``NIOAsyncWriter/yield(_:)``
+    /// will be resumed.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(_:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished(file:line:)`` error.
```
I don't think this is accurate.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,1060 @@
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will be resumed. Any subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will throw.
```
Same here re: yielding

FranzBusch replied:
I think this one is actually correct.

glbrntt commented on Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:
```diff
@@ -0,0 +1,1060 @@
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will be resumed. Any subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will throw.
```
Same here re: yielding

FranzBusch replied:
Same here

glbrntt reviewed (commented):
This is great @FranzBusch -- I think a couple of docs are out of date and should be fixed in a followup.
